### DISCUSSION TITLE
Multi-band reverse phase encoding

### DISCUSSION THREAD

**Discussion by Loxlan** [muscbridge/PyDKE#263] [Technical Support]
Great tool! A quick question, how would you use the tool for a multi-band reverse-phase encoding DWI dataset similar to the HCP dataset. 

 eg, sub-xx_acq-multiband_run-01_dwi.bval  sub-xx_acq-multiband_run-01_dwi.bvec  sub-xx_acq-multiband_run-01_dwi.json  sub-xx_acq-multiband_run-01_dwi.nii.gz

and reverse encoding as,
sub-xx_acq-multiband_run-02_dwi.bval  sub-xx_acq-multiband_run-02_dwi.bvec  sub-xx_acq-multiband_run-02_dwi.json  sub-xx_acq-multiband_run-02_dwi.nii.gz

Any example of how to process this data with singularity will be helpful.

Thanks

**Comment by TheJaeger** [comment]
Hi @Loxlan. You would parse all of those series into PyDesigner with the syntax

```
pydesigner -s -o [PATH TO OUTPUT]  sub-xx_acq-multiband_run-01_dwi.nii.gz,sub-xx_acq-multiband_run-01_dwi.nii.gz
```

**NOTE:** No spaces between commas separating paths to input files. 

This will allow PyDesigner to make its own internal decisions on achieving the highest possible SNR. Protocol information stored in JSON files is used to determine whether or not to run Gibbs unringing and figure out phase encoding information.

While we have not personally tested this with the HCP dataset, it should work if all relevant information is available in the JSON file. Running this in containers like Docker and Singularity will follow the same syntax with changes to file paths relative to mount point and/or other container control flags. An example on running PyDesigner within NeuroDock is available [here](https://pydesigner.readthedocs.io/en/latest/docker/docker_neurodock_run.html).

**Reply by Loxlan** [reply]
Thanks for the quick reply.

running using Singularity gave this error: 

ImportError: libgfortran.so.4: cannot open shared object file: No such file or directory

Not sure if this has to be installed in the Singularity image?

**Reply by TheJaeger** [reply]
Could you share your OS, Docker/Singularity version, PyDesigner version, and the full command you parsed?

**Reply by Loxlan** [reply]
OS-ubuntu 16.04, singularity 3.8 , neurodoc_v1.0-RC10 and the command,

singularity exec  -B /home/john/test_data/sub-001:/mnt -e /scratch/jhon/neurodock_v1.0-RC10.sif pydesigner --standard --output /mnt/processed /mnt/dwi/sub-001_acq-multiband_run-01_dwi.nii.gz,/mnt/dwi/sub-001_acq-multiband_run-02_dwi.nii.gz

I should add, this error is given only when running on the cluster.

**Reply by TheJaeger** [reply]
It seems that the Singularity image is not equipped with a library required to run one or more Python packages. Depending on the configuration of Singularity image (I believe it may be Ubuntu), you can add the following line to _neurodock_v1.0-RC10.sif_:

```
apt-get install gcc-7 g++-7 gfortran-7
```

Rebuild the image and see if it solves the issue.

**Reply by Loxlan** [reply]
Thanks, will give it a shot.

**Reply by Loxlan** [reply]
Got the container working. Just one think, I have a partial fourier data and wanted to use the -eddy_options  "--slm=linear", don't see how to specify this amoung the PyDesigner available options. Thanks

**Reply by TheJaeger** [reply]
PyDesigner does not allow users to specify any control flags for susceptibility induced distortion and eddy corrections. Control flags are internally handled to produce the most optimized image correction.

In your case, PyDesigner determines whether or not your input data set is fully shelled using the function [preprocessing.mrinfoutil.isfullsphere](https://github.com/m-ama/PyDesigner/blob/f948cf0585bd868ade5a8a241b88890a706999a3/designer/preprocessing/mrinfoutil.py#L446-L503) to toggle between `--slm=linear` or `--data_is_shelled` in [preprocessing.mrpreproc.undistort](https://github.com/m-ama/PyDesigner/blob/f948cf0585bd868ade5a8a241b88890a706999a3/designer/preprocessing/mrpreproc.py#L289-L421).

**Note:** Partial Fourier acquisitions affect Gibbs ringing correction rather than distortion correction, where the former is also automatically controlled by PyD.
